lib/logstash/outputs/elastic_workplace_search.rb

```diff
@@ -36,72 +40,96 @@ class LogStash::Outputs::ElasticWorkplaceSearch < LogStash::Outputs::Base
       batch = format_batch(events)
-      if @logger.trace?
-        @logger.trace("Sending bulk to Workplace Search", :size => batch.size, :data => batch.inspect)
-      end
+      @logger.trace('Sending bulk to Workplace Search', :size => batch.size, :data => batch.inspect) if @logger.trace?
       index(batch)
     end
   end
 
   private
+
+  def client_options
+    options = { :host => @url, :http_auth => @access_token.value }
+    options[:logger] = @logger if @logger.debug?
+    options[:tracer] = @logger if @logger.trace?
+    options
+  end
+
   def format_batch(events)
+    docs_per_source = {}
     events.map do |event|
       doc = event.to_hash
       # we need to remove default fields that start with "@"
       # since Elastic Workplace Search doesn't accept them
       if @timestamp_destination
-        doc[@timestamp_destination] = doc.delete("@timestamp")
+        doc[@timestamp_destination] = doc.delete('@timestamp')
       else # delete it
-        doc.delete("@timestamp")
+        doc.delete('@timestamp')
       end
-      if @document_id
-        doc["id"] = event.sprintf(@document_id)
+
+      doc['id'] = event.sprintf(@document_id) if @document_id
+      doc.delete('@version')
+
+      resolved_source = event.sprintf(@source)
+      unless docs_per_source[resolved_source]
+        @logger.debug('Creating new source segment in batch to send', resolved_source: resolved_source) if @logger.debug?
+        docs_per_source[resolved_source] = []
       end
-      doc.delete("@version")
-      doc
+      docs_per_source[resolved_source] << doc
     end
+
+    docs_per_source
   end
 
-  def index(documents)
-    response = @client.index_documents(@source, documents)
-    report(documents, response)
-  rescue => e
-    @logger.error("Failed to execute index operation. Retrying..", :exception => e.class, :reason => e.message)
-    sleep(1)
-    retry
+  def index(batch)
+    batch.each do |resolved_source, documents|
+      begin
+        if resolved_source =~ SOURCE_WITH_SPRINTF_REGEX || resolved_source =~ /^\s*$/
+          raise "Cannot resolve source field name #{@source} from event"
+        end
+
+        response = @client.index_documents(resolved_source, { :documents => documents })
+        report(documents, response)
+      rescue => e
+        @logger.error('Failed to execute index operation.', :exception => e.class, :reason => e.message,
+                      :resolved_source => resolved_source, :backtrace => e.backtrace, :retry => !@retry_disabled)
+
+        raise e if @retry_disabled
+
+        sleep(1)
+        retry
+      end
+    end
   end
 
   def report(documents, response)
     documents.each_with_index do |document, i|
-      errors = response["results"][i]["errors"]
-      if errors.empty?
-        @logger.trace? && @logger.trace("Document was indexed with no errors", :document => document)
+      errors = response.body&.dig('results', i, 'errors')
+      if errors&.empty?
```

Comment:
As I can understand 
```ruby
errors&.empty?
```

is `nil` if `errors` is `nil`. So it would print an error message reporting of a failed index.

If `errors` is nil, is it effectively a failed index operation?
In case `errors` is nil, also the log.warn would fail with an NPE because of `errors.to_a` is called on nil object

Reply:
AFAIK, `errors` should not be `nil` [swagger/v1/workplace-search.json](https://github.com/elastic/ent-search/blob/main/swagger/v1/workplace-search.json#L248), so this condition should never happen. Considering the `errors` array is the source of truth for confirming whether an indexing operation succeeded or failed, I've added it as a safeguard. 

I might be wrong, but having it `nil`, IMO, should be considered a failure as the API should not respond in that format.

Without this safeguard, it would raise an `undefined method` on the `errors.empty?`, which does not provide many details on what went wrong.

Reply:
If I'm not mistaken the `to_a` does not raise an NPE even if `errors` is `nil`. It parses it to empty array `[]` instead.

Reply:
Yes you are righ, `to_a` on a `nil` return empty array:

```sh
jruby-9.3.10.0 :003 > a = nil
 => nil
jruby-9.3.10.0 :004 > a.to_a
 => []
jruby-9.3.10.0 :005 >
```

So my concern, simply doesn't have meaning to exist.